docs/rfcs/003-consolidate-core-packages.md

```diff
@@ -0,0 +1,338 @@
+- They could read the CONTRIBUTING.md file which [has a section](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#atom-and-packages) that explains where to find the repos for core packages and how to contribute, but we don't really have a clear pointer to that in our README.md
+- If they don't happen to find that page, they might use Google to search for "atom tree view" and find the atom/tree-view repo and _maybe_ read the CONTRIBUTING.md file which sends them to Atom's overall contribution documentation
+- They might go to the Atom Forum or Slack community to ask how to contribute to a particular part of Atom and *hopefully* get a helpful response that points them in the right direction
+
+Having all of the core Atom packages represented in a top-level `packages` folder, even if they all don't actually live in the repo, will go a long way to making the core package code more discoverable.
+
+### Simpler core package contribution experience
+
+Separating core Atom features out into individual repositories and delivering them to Atom builds via `apm` is a great idea in theory because it validates the Atom package ecosystem and gives developers many examples of how to develop an Atom package.  It also gives Atom developers real-world experience working with Atom's APIs so that we ensure community package authors have the same hackability that Atom developers enjoy.
+
+On the other hand, having these packages live in separate repositories and released "independently" introduces a great deal of overhead when adding new features.  Here is a comparison of the current package development workflow contrasted to what we could achieve with consolidated packages:
+
+#### Current Package Development Workflow
+
+For example, to add a single feature to the `tree-view` package, one must:
+
+1. Fork and clone the `tree-view` repository to their computer (making sure to pull the commit relevant to the version of Atom they are working with)
+1. Run `apm install` and `apm link` inside of the repo folder
+1. Make their desired changes to the code
+1. Open a PR to the `tree-view` repo and wait for CI to pass and a maintainer to review it
+1. Work with maintainers to get the PR approved and merged
+
+After this is finished, an Atom maintainer must take the following steps:
+
+1. Clone the `tree-view` repo
+2. Run `apm publish` to publish a new release of the package
+3. Edit `package.json` in the Atom repo to reflect the new version of `tree-view`
+4. Commit and push the changes to the relevant branch where the change belongs (`master` or `1.nn-releases`)
+
+#### Simplified Package Development
+
+If we were to move `tree-view` (or any other core Atom package) back into `atom/atom`, the development workflow would look more like this:
+
+1. Fork and clone `atom/atom` and switch to a release branch if necessary
+1. Build Atom and launch it in dev mode
+1. Make desired changes to the code in `packages/tree-view`
```

Comment:
While you're developing, how easy is it to play with your changes in a live environment? Do you need to do a full `atom/atom` rebuild, or just launch in dev mode? What about if you change dependencies?

Reply:
Just launch in dev mode!  However, I think you found the first major issue with how I implemented PR #17686: the folders under `packages` never have `apm install` run directly on them so their `node_modules` folders are not populated by the time they're loaded in dev mode.

Asking the user to go run `apm install` inside the package folders they want to work on sounds like a lot of extra friction so I'll have to think about this a bit.  Worst case scenario we can use the original strategy I started with, running `apm install` on each package folder individually and then copying those folders into Atom's`node_modules` during the full build, possibly running `npm dedupe` afterward.